# Modelo de Previsão de Sexo do Pacientes

## Test - Data Scientist (profile statistician)

### Amanda Cristina da Costa Guimaraes

In [1]:
!pip install graphviz
!graphviz

'graphviz' não é reconhecido como um comando interno
ou externo, um programa operável ou um arquivo em lotes.


In [2]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_csv('test_data_CANDIDATE.csv')

Vamos ler o arquivo csv e transformar em um dataframe

In [4]:
df['sex'] = df['sex'].replace('M', 0)

In [5]:
df['sex'] = df['sex'].replace('m', 0)

In [6]:
df['sex'] = df['sex'].replace('F', 1)

In [7]:
df['sex'] = df['sex'].replace('f', 1)

Vamos converter os dados de sexo para 0 e 1 sendo:
    - masculino (M) = 0
    - feminino (F) = 1

In [8]:
df['sex'].value_counts()

1    196
0     92
Name: sex, dtype: int64

Após a transformação dos dados temos em nosso dataframe:
- 92 pacientes do sexo masculino 
- 196 pacientes do sexo feminino

Vamos define os valores do dataframe "X" e "Y" sendo:
- X = Cor do cabelo (hc) e Cor da pele (sk)
- Y = Sexo (sex)

In [9]:
x = df[['hc', 'sk']] 
y = df[['sex']]

Verificando como está o x.

In [10]:
x.head()

hc  sk
0   0   1
1   1   3
2   0   1
3   1   0
4   1   0

Verificando como está o y.

In [11]:
y.head()

sex
0    0
1    1
2    1
3    1
4    1

### Para analisar essas classes eu vou precisar de aloritimos estimadores, e para isso existem as bibliotecas do Python como: sklearn

Queremos criar um mecanismo que consiga estimar a classe desse tipo de dado. Para isso, usaremos algorítimos de machine learning, data analysis e estatística que estão armazenados nas diversas bibliotecas do Python. A que usaremos se chama Scikit Learn.

Do módulo sklearn.svm, importaremos um estimador chamado LinearSVC. Posteriomente, entraremos em mais detalhes sobre alguns estimadores existentes e o que cada um é capaz de realizar. Por enquanto, usaremos esse estimador básico.

In [12]:
from sklearn.svm import LinearSVC
model = LinearSVC()

In [13]:
treino_x = x[:75]
treino_y = y[:75]

### Treina a IA

In [14]:
from sklearn.svm import LinearSVC

In [15]:
from sklearn.metrics import accuracy_score

In [16]:
treino_x = x[:75]
treino_y = y[:75]
teste_x = x[75:]
teste_y = y[75:]
teste_y.shape

print("Treinaremos com %d elementos e testaremos com %d elementos" % (len(treino_x), len(teste_x)))

Treinaremos com 75 elementos e testaremos com 213 elementos


Primeiramente, com df.shape verificaremos quantos elementos temos nos dados e o formato deles:

In [17]:
df.shape

(288, 18)

Veremos que em nosso arquivo há 288 linhas e 18 colunas. Separaremos em média 25% para testar o algorítimo, e o restante (cerca de 75% dos dados) para o treinamento. Portanto, para treino_x, coletaremos os primeiros 75 elementos (treino_x = x[:75]). Podemos utilizar treino_x.shape para verificar se o número de elementos está de fato correto:

In [18]:
treino_x = x[:75]
treino_y = y[:75]

Para confirmarmos se as matrizes estão com a quantidade correta de elementos, acionaremos teste_y.shape. Por fim, registraremos essas informações imprimindo o tamanho (len()) de treino_x e treino_y utilizando print().

In [19]:
treino_x = x[:75]
treino_y = y[:75]
teste_x = x[75:]
teste_y = y[75:]
teste_y.shape

print("Treinaremos com %d elementos e testaremos com %d elementos" % (len(treino_x), len(teste_x)))

Treinaremos com 75 elementos e testaremos com 213 elementos


Para treinarmos e executarmos o algorítimo usaremos a motodologia  do sklearn.svm importaremos LinearSVC, e treinaremos o modelo com os dados treino_x e treino_y.

In [20]:
from sklearn.svm import LinearSVC

modelo = LinearSVC()
modelo.fit(treino_x, treino_y)

C:\Users\Amanda\anaconda3\lib\site-packages\sklearn\utils\validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LinearSVC()

In [21]:
from sklearn.svm import LinearSVC
from sklearn.metrics import accuracy_score

modelo = LinearSVC()
modelo.fit(treino_x, treino_y)
previsoes = modelo.predict(teste_x)

acuracia = accuracy_score(teste_y, previsoes) * 100
print("A acurácia foi %.2f%%" % acuracia)

A acurácia foi 63.38%


C:\Users\Amanda\anaconda3\lib\site-packages\sklearn\utils\validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


### Testa o Modelo

In [22]:
previsões = modelo.predict(teste_x)